Fix overflow_check

gsnedders opened the pull request:
Fixes #3588, and while we're at it improves all the (arithmetic) overflow detection code, including adding support for the gcc 5+/clang 3.4+ `__builtin_XXX_overflow`.

Looking at `Demos/benchmarks/chaos.py` (the only file in that directory significantly affected by enabling `overflowcheck`, with the non-`__builtin_XXX_overflow` code paths we gain 2%, and with them we gain 4% (and that's then only 1% behind `overflowcheck=False`).

gsnedders committed:
Fix #3588: Make existing overflow code safe

Signed overflow is undefined behaviour in C and compilers can and do
optimized on that basis.

gsnedders committed:
Use __builtin_XXX_overflow if available

This is much quicker in general, as these all just then read the
overflow flag from the status register

gsnedders committed:
Speed up our overflow impls

Note this is primarily based on performance of compilers which do not
support __builtin_add_overflow (i.e., not Clang >= 3.4 & gcc >= 5),
mostly looking at several gcc 4 releases (used by older, supported,
RHEL releases and Debian 8 "Jessie") and MSVC.

Implementations of all of these functions using the builtins where
available is forthcoming.

da-woods commented on Cython/Utility/Overflow.c:
```diff
@@ -46,6 +46,22 @@ static int __Pyx_check_twos_complement(void) {
 
+#if defined(__has_builtin)
+#  if __has_builtin(__builtin_add_overflow) && !defined(__ibmxl__)
+#    define HAVE_BUILTIN_OVERFLOW
```
I think most macros like this start with `CYTHON_` (e.g. `CYTHON_HAVE_BUILTIN_OVERFLOW`), just to reduce the chances of conflicting with someone else's macro.

scoder commented on Cython/Utility/Overflow.c:
```diff
@@ -46,6 +46,14 @@ static int __Pyx_check_twos_complement(void) {
 #define __Pyx_div_const_no_overflow(a, b, overflow) ((a) / (b))
 
+#if defined(__GNUC__)
+#  define __Pyx_constant_p(x) (__builtin_constant_p(x))
```
`__Pyx_is_constant(x)` ?

scoder commented on Cython/Utility/Overflow.c:
```diff
@@ -95,20 +108,28 @@ static CYTHON_INLINE {{UINT}} __Pyx_mul_{{NAME}}_checking_overflow({{UINT}} a, {
         return r;
 #endif
     } else {
-        {{UINT}} prod = a * b;
-        double dprod = ((double) a) * ((double) b);
+        // somewhat surprisingly, at least on x86_64 it's quicker to use floats (c.f. the const
+        // variant below)
```
Ok, but – is the value range large enough to cover an `unsigned long long` multiplication?

gsnedders replied:
Looking at this much more closely, this is actually unsafe when `unsigned long long` is 64-bit (the original intent of this file was to allow false-positive overflow reporting, but not false-negatives, but I realise elsewhere we always raise OverflowError if this is truthy) unless we tightly control rounding modes due to 2^64 - 1 not being a possible value (and we probably don't want to get into changing rounding mode because that's a world of pain!), which is likewise true with a double.

I'm also doubting myself as to whether this is actually quicker v. the const version for the unsigned case anyway…

scoder commented on Cython/Utility/Overflow.c:
```diff
@@ -95,20 +108,28 @@ static CYTHON_INLINE {{UINT}} __Pyx_mul_{{NAME}}_checking_overflow({{UINT}} a, {
-    if (b > 1) {
-        *overflow |= a > __PYX_MAX({{UINT}}) / b;
+    if (__Pyx_constant_p(a) && !__Pyx_constant_p(b)) {
+        // paranoia
```
I'm aware that there's a longer comment in the signed variant further down, but it seems worth repeating it here. For a future reader, it will be less obvious what the 'paranoia' refers to. (I'm ok with keeping this safety check in general.)

scoder commented on Cython/Utility/Overflow.c:
```diff
@@ -170,17 +170,17 @@ static CYTHON_INLINE {{INT}} __Pyx_add_const_{{NAME}}_checking_overflow({{INT}}
         *overflow |= a < __PYX_MIN({{INT}}) - b;
     }
-    return a + b;
+    return ({{INT}}) ((unsigned {{INT}}) a + (unsigned {{INT}}) b);
```
Could you at least add a comment here? It's not obvious from seeing the code that this is doing something reasonable.

I think referring to the SO question might be ok:
https://stackoverflow.com/questions/4240748/allowing-signed-integer-overflows-in-c-c

scoder reviewed (commented):
Nice! I left some comments, mostly about readability. This code is so deeply technical that it's worth dropping a comment more here and there.

gsnedders reviewed (commented):
I did write some comments, but wasn't sure quite how far to go given I'd spent long enough staring at the code, so did partly put this up as it was just to see where you wanted more! Looking at this code again (I wrote most of it a while ago) I can't say I disagree.

gsnedders committed:
fixup! Use __builtin_XXX_overflow if available

gsnedders committed:
fixup! Speed up our overflow impls

gsnedders committed:
fixup! Speed up our overflow impls

gsnedders committed:
fixup! Speed up our overflow impls

Reviewing this closely, the only way this can be correct for 64-bit
ints and above is when we carefully manage floating point rounding
mode, and that's vastly more pain (and the standard ways to change
rounding mode inhibit compiler optimisations much more broadly).

gsnedders committed:
fixup! Speed up our overflow impls

gsnedders committed:
fixup! Speed up our overflow impls

gsnedders committed:
Ensure lshift doesn't hit undefined-behaviour

gsnedders commented:
@scoder ping

scoder commented:
Thanks!
I'll target this to 3.0 for now, mostly to give it some more testing. Wouldn't mind backporting it to 0.29.x at some point.
